### PR TITLE
SOLR-15399: IndexFetcher should not issue a local commit for PULL replicas

### PR DESCRIPTION
**Pull request by thelabdude**
# Description

When doing the work for https://github.com/apache/solr/commit/ae1ac22eb70d44f80e5c89241bdd0d3e4144e62a, I discovered that the `TestPullReplica` class was disabled via an AwaitsFix. I enabled it and it passed fairly consistently, but then saw flaky failures on Jenkins. This PR fixes those flaky tests ...

`IndexFetcher` was triggering a local commit for a PULL replica core (see snippet below) which caused a race condition where the PULL replica would think its index generation / version were the same as the leader (see log I posted in JIRA). 

From IndexFetcher, line 479:
```
          if (skipCommitOnLeaderVersionZero) {
            openNewSearcherAndUpdateCommitPoint();
          } else {
            SolrQueryRequest req = new LocalSolrQueryRequest(solrCore, new ModifiableSolrParams());
            solrCore.getUpdateHandler().commit(new CommitUpdateCommand(req, false));
          }
```
(the else block hitting is what causes the tests to fail intermittently)

This would cause the PULL replica to not sync index updates with the leader in the tests (since they do it doc at a time).

# Solution

I don't think PULL replicas should ever commit locally and should always get the index from the leader. So this PR ensures that `skipCommitOnLeaderVersionZero` is `true` for PULL replicas, which avoids this race condition.

# Tests

Removing `@BadApple` from several flaky tests that pass consistently now: 200/200 beasts (tests would fail within about 40 beast runs w/o this fix)

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Review by tflobbe — APPROVED** [review]
Thanks for looking into this, Tim. LGTM. I'm guessing the side effects are that the PULL replica won't really clean up it's index (i.e. disk will still contain all index files) until a real-non-gen-zero replication happen from the leader, right? That index won't be searchable though. Even if there is a restart, since the PULL replica will go through the recovery process, the index on disk won't be searchable again IIUC. This should also solve SOLR-10751 and SOLR-12100

**Comment by thelabdude** [comment]
thanks for the review @tflobbe ... let me add a test case for the restart while the PULL replica is in this side-effect state, just so we can be sure it's ok and recovers properly

File: solr/core/src/java/org/apache/solr/cloud/ReplicateFromLeader.java [hunk 1]
```diff
@@ -87,7 +88,20 @@ public void startReplication(boolean switchTransactionLog) {
 
       NamedList<Object> followerConfig = new NamedList<>();
       followerConfig.add("fetchFromLeader", Boolean.TRUE);
-      followerConfig.add(ReplicationHandler.SKIP_COMMIT_ON_LEADER_VERSION_ZERO, switchTransactionLog);
+
+      // don't commit on leader version zero for PULL replicas as PULL should only get its index state from leader
+      boolean skipCommitOnLeaderVersionZero = switchTransactionLog;
+      final CloudDescriptor cloudDescriptor = core.getCoreDescriptor().getCloudDescriptor();
+      if (cloudDescriptor != null) {
```
Comment (HoustonPutman) [inline solr/core/src/java/org/apache/solr/cloud/ReplicateFromLeader.java:95]: maybe only check this if !skipCommitOnLeaderVersionZero

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java [hunk 2]
```diff
@@ -258,6 +260,37 @@ public void testAddDocs() throws Exception {
...
+    for (Replica r : pullReplicas) {
+      JettySolrRunner jetty = cluster.getReplicaJetty(r);
+      if (!jetty.getNodeName().equals(leaderJetty.getNodeName())) {
+        replicaJetty = jetty;
+        break;
+      }
+    }
+
+    // stop / start the node with a pull replica
+    if (replicaJetty != null) {
```
[the remark this answers is not in the document]
Comment (TomMD) [reply solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:280]: @muse-dev ignore

I see this false positive bug popping up all of a sudden.  Will investigate.

[the remark this answers is not in the document]
Comment (thelabdude) [reply solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:280]: Thanks @TomMD  ... couldn't figure out what it was complaining about there ;-)

**Comment by dsmiley** [comment]
Impressive work to track this down!

**Pull request merged**